README.md
# **Kinetic-py**
[![Travis](https://img.shields.io/travis/Seagate/kinetic-py.svg)](https://travis-ci.org/Seagate/kinetic-py)
[![PyPI](https://img.shields.io/pypi/v/kinetic.svg)](https://pypi.python.org/pypi/kinetic/)
[![PyPI](https://img.shields.io/pypi/status/kinetic.svg)](https://pypi.python.org/pypi/kinetic/)
[![PyPI](https://img.shields.io/pypi/l/kinetic.svg)](https://github.com/Seagate/kinetic-py/blob/master/LICENSE/LGPL2.1.md)

Introduction
============
The [kinetic-protocol](https://github.com/Seagate/kinetic-protocol) python client.

## Requirements
- Requires Python 2.7.3 or higher. 
- Requires Python 2.7.9 on OSX to use SSL 

> **NOTE:** Python 3.x is not supported. 
 
Installing latest stable release
================================
    pip install kinetic


Installing from Source
======================

    git clone https://github.com/Seagate/kinetic-py.git
    cd kinetic-py
    python setup.py develop

> **NOTE:** for devices with old firmware code get version 0.7.3 of the libray
    git checkout 0.7.3

Running Tests
=============
The tests need a Kinetic device to run. You can use the simulator available at https://github.com/Seagate/kinetic-java.
To configure the test environment:

    export KINETIC_HOST=192.168.0.20
    export KINETIC_PORT=8123

Optionally you can point the tests to the simulator jar:

    export KINETIC_JAR=kinetic-simulator-<currentversion>-jar-with-dependencies.jar

Then to run the tests:

    python test/

Getting Started with the basic client
=====================================

```python
from kinetic import Client
c = Client('localhost', 8123)
c.connect()
c.put('message','hello world')
print c.get('message').value
```
Should print out _hello world_

Troubleshooting during the installation
=======================================
On a brand new system, you might be missing a few things.
If you get an error saying setup tools not installed or missing.
Check the python [setuptools intallation guide](https://pypi.python.org/pypi/setuptools#installation-instructions).
If you needed to installed that, chances are you are missing some requirements to install and compile eventlet on your system.
On debian systems the quickest way is `sudo apt-get install python-eventlet`.


License
-------

This project is licensed under GNU Lesser General Public License, version 2.1
* [Markdown](LICENSE/LGPL2.1.md) version
* [Original](LICENSE/LGPL2.1.txt) version


CHANGES.md
Changes since 0.9.1
===========================
>**NOTE:** this section will document changes to the library since the last release

## Minor changes
- Added `on_faulted` handler on `Client`. 
- Added `operation_timeout` support on `Client` for **blocking** calls.

Changes from 0.9.0 to 0.9.1
===========================
This version was skipped due to a pypi error.

Changes from 0.8.2 to 0.9.0
===========================

## Important
- Kinetic Protocol version updated to [3.0.6](https://github.com/Seagate/kinetic-protocol/tree/3.0.6)
- Deprecrated classes and methods will be removed on the next major release.

## New features
- Added support for Batch operations
- Added `reverse` parameter on GetKeyRange operation 

## Major changes
- `AsyncClient` has been renamed to `Client`
- A new `SecureClient` has been added to `kinetic`.

## Minor changes
- Added env variable _KINETIC_CONNECT_TIMEOUT_ to control default connection timeout.

## Deprecated features
- Old blocking `Client` has been moved to `kinetic.depracated.BlockingClient`
- Old `AdminClient` has been moved to `kinetic.depracated.AdminClient`

## Misc
- Added alias for `AsyncClient = Client` to smooth transition.
- Added alias on `kinetic` for `AdminClient` to smooth transition.

Changes from 0.8.1 to 0.8.2
===========================

## Bug fixes
- Keys with empty values show correctly as '' instead of None
- **WRITEBACK** set as default mode for `put` and `delete` operations

## Misc
- OSX requirements for SSL connections updated.

Changes from 0.8.0 to 0.8.1
===========================
This section will document changes to the library since the last release

## Important
- Kinetic Protocol version updated to [3.0.5](https://github.com/Seagate/kinetic-protocol/tree/3.0.5)

## New features
- Added timeout, time_quanta, priority and early_exit support
- Pin size added to limits
- SHA1 calculation used as default on puts when algorithm and tag not specified.

## Minor changes
- The client will use WRITEBACK as default when synchronization mode not specified on PUT and DELETE.

Changes from 0.7.3 to 0.8.0
===========================

## Important
- Kinetic Protocol version updated to [3.0.0](https://github.com/Seagate/kinetic-protocol/tree/3.0.0)
- Everything requires proto 3.0.0 or higher on the device
- Devices pre 3.0.0 do not have handshakes and will raise Handshake timeout

## New features
- Added `--version` to cmd line tool
- Added background operations Scan and Optimize
- Added pin operations (Lock, unlock, erase, secureErase)
- Client auto configures cluster_version based on initial handshake
- ErasePin and LockPin can be set during the security operation
- Client fields config and limits show device information
- Added unsolicited status support
- Pin based operations MUST have a pin set and SSL enabled
- setSecurity MUST have SSL enabled
- getLog / getLogAsync added to Client/AsyncClient

## Behavior changes
- Removed GreenClient (Feature overlap with AsyncClient)
- Removed PipelinedClient (Only used internally by the kineticc)
- setPin removed from AdminClient

Changes from 0.7.2 to 0.7.3
===========================

## Important
- Kinetic Protocol version updated to [2.0.6](https://github.com/Seagate/kinetic-protocol/tree/2.0.6)

## New features
- Added handshake to negotiate connection id during connect
- Added TCP_NODELAY socket option to improve performance
- GetKeyRange operations now accept None arguments on keys (Issue #10)

## Behavior changes
- Removed DEVICE log from LogTypes.add() (Issue #15)

## Bug Fixes
- Fixed problem with status validation on the AdminClient
- Fixed error message when Magic number is invalid

Changes from 0.7.1 to 0.7.2
===========================

## Important
- Kinetic Protocol version updated to [2.0.5](https://github.com/Seagate/kinetic-protocol/tree/2.0.5)
- The compiled python proto kinetic/kinetic_pb2.py is now included on the repo

## New features
- Added zero copy support on puts and gets (Requires splice system call)
- Added IPv6 address support on all clients (Issue #8)
- Added new exception type ClusterVersionFailureException (Requires drive version 2.0.4)
- Added new Device specific GetLog (Requires protocol 2.0.5)
- Added SSL/TLS support on all clients

## Bug Fixes
- Fixed bug on invalid magix number (PR #11 contributed by @rpcope1, ASOKVAD-313)
- Fixex bug that caused close() and connect() to fail on a connection that faulted (Issue #7)
- Fixed a bug that caused the AsyncClient to crash when calling close()

Changes from 0.7.0 to 0.7.1
===========================

## New features
- Added setSecurity on AdminClient
- Added getVersion and getVersionAsync to the library.

## Bug Fixes
- Fixed tests not running and testcases with hardcoded 'localhost'
- Fixed flush operation build parameters (Merge #5, contributed by @rpcope1).
- AsyncClient returns NotConnected exception when an operation is attempted on a client before calling connect().
- Lowered default number of keys asked on ranges to 200 (ASKOVAD-287).
- Fixed typo on baset test case (Merge #2, contributed by @zaitcev).

Changes from 0.6.0.2 to 0.7.0
=============================

## Important
Kinetic Protocol version updated to 2.0.3

## New features
- Added Flush command (Requires protocol 2.0.3).
- Added Limits section on GetLog (Requires protocol 2.0.3).
- Added protocol_version field on kinetic module.
- Added version field on kinetic module.

## Breaking changes
- Renamed Synchronization.ASYNC to Synchronization.WRITEBACK
- Renamed Synchronization.SYNC to Synchronization.WRITETHROUGH

## Bug Fixes
- Fixed issue with asynchronous clients leaving the socket open after _close_ was called (ASOKVAD-263).


kinetic/common.py
# Copyright (C) 2014 Seagate Technology.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

#@author: Ignacio Corderi

import kinetic_pb2 as messages
import eventlet
import os

MAX_KEY_SIZE = 4*1024
MAX_VALUE_SIZE = 1024*1024

DEFAULT_CONNECT_TIMEOUT = 0.1
DEFAULT_SOCKET_TIMEOUT = 5
DEFAULT_CHUNK_SIZE = 64*1024

# Env variables
try: 
    KINETIC_CONNECT_TIMEOUT = float(os.environ.get('KINETIC_CONNECT_TIMEOUT', DEFAULT_CONNECT_TIMEOUT))
except:
    KINETIC_CONNECT_TIMEOUT = DEFAULT_CONNECT_TIMEOUT
    
local = messages.Local()

class DeferedValue():

    def __init__(self, socket, value_ln):
        self.socket = socket
        self.length = value_ln
        self._evt = eventlet.event.Event()

    def set(self):
        self._evt.send()

    def wait(self):
        self._evt.wait()


class Entry(object):

    #RPC: Note, you could build this as a class method, if you wanted the fromMessage to build
    #the subclass on a fromMessage. I suspect you always want to generate Entry objects,
    #in which case, a staticmethod is appropriate as a factory.
    @staticmethod
    def fromMessage(command, value):
        if not command: return None
        return Entry(command.body.keyValue.key, value, EntryMetadata.fromMessage(command))

    @staticmethod
    def fromResponse(response, value):
        if (response.status.code == messages.Command.Status.SUCCESS):
            return Entry.fromMessage(response, value)
        elif (response.status.code == messages.Command.Status.NOT_FOUND):
            return None
        else:
            raise KineticClientException("Invalid response status, can' build entry from response.")

    def __init__(self, key, value, metadata=None):
        self.key = key
        self.value = value
        self.metadata = metadata or EntryMetadata()

    def __str__(self):
        if self.value:
            return "{key}={value}".format(key=self.key, value=self.value)
        else:
            return self.key


class EntryMetadata(object):

    @staticmethod
    def fromMessage(command):
        if not command: return None
        return EntryMetadata(command.body.keyValue.dbVersion, command.body.keyValue.tag,
                             command.body.keyValue.algorithm)

    def __init__(self, version=None, tag=None, algorithm=None):
        self.version = version
        self.tag = tag
        self.algorithm = algorithm

    def __str__(self):
        return self.version or "None"


class KeyRange(object):

    def __init__(self, startKey, endKey, startKeyInclusive=True,
                 endKeyInclusive=True):
        self.startKey = startKey
        self.endKey = endKey
        self.startKeyInclusive = startKeyInclusive
        self.endKeyInclusive = endKeyInclusive

    def getFrom(self, client, max=1024):
        return client.getKeyRange(self.startKey, self.endKey, self.startKeyInclusive, self.endKeyInclusive, max)


class P2pOp(object):

    def __init__(self, key, version=None, newKey=None, force=None):
        self.key = key
        self.version = version
        self.newKey = newKey
        self.force = force


class Peer(object):

    def __init__(self, hostname='localhost', port=8123, tls=None):
        self.hostname = hostname
        self.port = port
        self.tls = tls

# Exceptions

class KineticException(Exception):

    def __init__(self, value):
        self.value = value

    def __str__(self):
        return repr(self.value)

class KineticClientException(KineticException):
    pass
    
class OperationTimeout(KineticException):
    pass    

class NotConnected(KineticClientException):
    pass

class AlreadyConnected(KineticClientException):
    pass

class ServerDisconnect(KineticClientException):
    pass

class ConnectionFaulted(KineticClientException):
    pass

class ConnectionClosed(KineticClientException):
    pass

class KineticMessageException(KineticException):

    def __init__(self, status):
        self.value = status.statusMessage
        self.status = status
        self.code = self.status.DESCRIPTOR.enum_types[0]\
                .values_by_number[self.status.code].name

    def __str__(self):
        return self.code + (': %s' % self.value if self.value else '')

class ClusterVersionFailureException(KineticMessageException):

    def __init__(self, status, cluster_version):
        super(ClusterVersionFailureException, self).__init__(status)
        self.cluster_version = cluster_version

class BatchAbortedException(KineticException):
    def __init__(self, value):
        super(BatchAbortedException, self).__init__(value)
        self.failed_operation_index = -1

class BatchCompletedException(KineticClientException):
    def __init__(self):
        super(BatchCompletedException, self).__init__('batch completed. no more operations are permitted within this batch.')

class HmacAlgorithms:
    INVALID_HMAC_ALGORITHM = -1 # Must come first, so default is invalid
    HmacSHA1 = 1 # this is the default


class Priority:
    NORMAL = 5
    LOWEST = 1
    LOWER = 3
    HIGHER = 7
    HIGHEST = 9


class ACL(object):

    DEFAULT_IDENTITY=1
    DEFAULT_KEY = "asdfasdf"

    def __init__(self, identity=DEFAULT_IDENTITY, key=DEFAULT_KEY, algorithm=HmacAlgorithms.HmacSHA1, max_priority=Priority.NORMAL):
        self.identity = identity
        self.key = key
        self.hmacAlgorithm = algorithm
        self.domains = set()
        self.max_priority = max_priority


class Domain(object):
    """
        Domain object, which corresponds to the domain object in the Java client,
        and is the Scope object in the protobuf.
    """
    def __init__(self, roles=None, tlsRequried=False, offset=None, value=None):
        if roles:
            self.roles = set(roles)
        else:
            self.roles = set()
        self.tlsRequired = tlsRequried
        self.offset = offset
        self.value = value


class Roles(object):
    """
        Role enumeration, which is the same thing as the permission field for each
        scope in the protobuf ACL list.
    """
    READ = 0
    WRITE = 1
    DELETE = 2
    RANGE = 3
    SETUP = 4
    P2POP = 5
    GETLOG = 7
    SECURITY = 8

    @classmethod
    def all(cls):
        """
            Return the set of all possible roles.
        """
        return [cls.READ, cls.WRITE, cls.DELETE, cls.RANGE, cls.SETUP, cls.P2POP, cls.GETLOG, cls.SECURITY]


class Synchronization:
    INVALID_SYNCHRONIZATION = -1
    WRITETHROUGH = 1 # Sync
    WRITEBACK = 2 # Async
    FLUSH = 3


class IntegrityAlgorithms:
    SHA1 = 1
    SHA2 = 2
    SHA3 = 3
    CRC32 = 4
    CRC64 = 5
    # 6-99 are reserverd.
    # 100-inf are private algorithms


class LogTypes:
    INVALID_TYPE = -1
    UTILIZATIONS = 0
    TEMPERATURES = 1
    CAPACITIES = 2
    CONFIGURATION = 3
    STATISTICS = 4
    MESSAGES = 5
    LIMITS = 6
    DEVICE = 7

    @classmethod
    def all(cls):
        """
            LogTypes.all takes no arguments and returns a list of all valid log magic numbers (from the protobuf definition)
            that can be retrieved using the AdminClient .getLog method. Log types avaiable are: (0-> Utilizations, 1-> Temperatures,
            2->Drive Capacity, 3-> Drive Configuration, 4->Drive usage statistics, and 5-> Drive messages). This can be passed as
            the sole argument to the AdminClient.getLog function.
        """
        return [cls.UTILIZATIONS, cls.TEMPERATURES, cls.CAPACITIES, cls.CONFIGURATION, cls.STATISTICS, cls.MESSAGES,
                cls.LIMITS]
